bittide/src/Bittide/Transceiver.hs

```diff
@@ -7,453 +7,550 @@
+--  2. Check PRBS data
+--  3. Signal that PRBS data is OK after observing it for some time (see section
+--    __Reset manager__).
+--  4. Wait for 'Meta.lastPrbsWord'
+--  5. Freeze alignment logic
+--
 module Bittide.Transceiver where
 
 import Clash.Explicit.Prelude
-import Clash.Explicit.Reset.Extra
-import Clash.Cores.Xilinx.GTH
-import Clash.Cores.Xilinx.Xpm.Cdc.Single
 
-import Bittide.Arithmetic.Time
+import Bittide.Arithmetic.Time (Milliseconds, trueForSteps)
+import Bittide.ElasticBuffer (sticky)
+import Clash.Cores.Xilinx.GTH (gthCore)
+import Clash.Cores.Xilinx.Ila (IlaConfig(advancedTriggers, depth, stages), ilaConfig, ila, Depth(D1024))
+import Clash.Cores.Xilinx.Xpm.Cdc.ArraySingle (xpmCdcArraySingle)
+import Clash.Cores.Xilinx.Xpm.Cdc.Single (xpmCdcSingle)
+import Clash.Explicit.Reset.Extra (Asserted(Asserted), delayReset, xpmResetSynchronizer)
+import Clash.Prelude (withClock)
+import Clash.Sized.Vector.Extra (zipWith8)
+import Control.Monad (when)
+import Data.Maybe (isNothing, fromMaybe)
+import Data.Proxy (Proxy(Proxy))
+
+import qualified Bittide.Transceiver.Cdc as Cdc
+import qualified Bittide.Transceiver.Comma as Comma
+import qualified Bittide.Transceiver.Prbs as Prbs
+import qualified Bittide.Transceiver.ResetManager as ResetManager
+import qualified Bittide.Transceiver.WordAlign as WordAlign
+
+-- | Meta information send along with the PRBS and alignment symbols. See module
+-- documentation for more information.
+data Meta = Meta
+  { ready :: Bool
+  -- ^ Ready to receive user data
+  , lastPrbsWord :: Bool
+  -- ^ Next word will be user data
+  , fpgaIndex :: Unsigned 3
+  -- ^ FPGA index to use for debug signals
+  , transceiverIndex :: Unsigned 3
+  -- ^ Transceiver index to use for debug signals
+  }
+  deriving (Generic, NFDataX, BitPack)
+
+-- | Insert zeroes such that each of the following are encoded in 4 bits, making
+-- them easier to read when formatted as a hex value:
+--
+--   * prbsOk, lastPrbsWord
+--   * fpgaIndex
+--   * transceiverIndex
+--
+-- This is useful for when we don't control formatting (such as when looking at
+-- ILA traces).
+prettifyMetaBits :: BitVector 8 -> BitVector 12
+prettifyMetaBits bv = pack $
+  let meta = unpack @Meta bv in
+  ( low, low, meta.ready
+  , meta.lastPrbsWord
+  , low, meta.fpgaIndex
+  , low, meta.transceiverIndex
+  )
+
+data Config dom = Config
+  { debugIla :: Bool
+  -- ^ Instantiate a debug ILAs
+  , debugFpgaIndex :: Signal dom (Unsigned 3)
+  -- ^ FPGA index to use for debug signals
+  , resetManagerConfig :: ResetManager.Config
+  -- ^ Configuration for 'ResetManager.resetManager'
+  }
+
+defConfig :: Config dom
+defConfig = Config
+  { debugIla = False
+  , debugFpgaIndex = pure 0
+  , resetManagerConfig = ResetManager.defConfig
+  }
 
 -- | Careful: the domains for each transceiver are different, even if their
 -- types say otherwise.
-data TransceiverOutputs n tx rx txS freeclk = TransceiverOutputs
+data Outputs n tx rx txS free = Outputs
```

Comment:
Is there a good reasons to package `Input` and `Output` vectors this way? It looks to me like it would just introduce conversion overhead for switching between the two representations.

Reply:
I'll let you be the judge of whether it's a _good_ reason, but the idea is that one can do something like:

```haskell
outputs = transceiversPrbsN [..]

allUp = and <$> outputs.linkUps
```

..instead of having to manually pick out the `linkUp` signals from vectors.

Reply:
I don't see any particular overhead in using the following
```haskell
  transceivers =
    transceiverPrbs
      @GthTx @GthRx @Ext200 @Basic125 @GthTx @GthRx
      Gth.gthCore defConfig{debugIla=True, debugFpgaIndex=bitCoerce <$> fpgaIndex}
      <$> ( Input sysClk gthAllReset refClk
              <$> iterateI (+1) 0
              <*> clockPaths
              <*> channelNames
              <*> unbundle (unpack <$> rxNs)
              <*> unbundle (unpack <$> rxPs)
              <*> counters
              <*> repeat (pure True)
              <*> repeat (pure True)
          )

  allUp = and . linkUp <$> transceivers
```
compared to
https://github.com/bittide/bittide-hardware/blob/e27f60e3fd216f0643f09dbbe4acdb81f9a5f770/bittide-instances/src/Bittide/Instances/Hitl/Transceivers.hs#L147-L162 + `allUp = and <$> outputs.linkUps`

except that you can't use record syntax in the former case. However, if this is all just for introducing names then it definitely is boilerplate overhead. Especially since you enable `NoFieldSelectors`, which means using for example `linkUp` from other module as a function requires nasty re-introductions, like
```haskell
  linkUp' :: Output a b c dom d -> Signal dom Bool
  linkUp' x = x.linkUp
```
---
We shouldn't introduce tons of definitions just to support a particular coding style. Remember that others may like to use other styles and this way of implementation looks all quite restrictive and cumbersome to me in that sense.

Reply:
> except that you can't use record syntax in the former case. However, if this is all just for introducing names then it definitely is boilerplate overhead.

Yes, this is all for introducing names! Positional arguments aren't that great, especially when there is many of them. It's easy to swap arguments (Does P come before N? Does clock path come before channel name?), you can't see at the instantiation site what the argument names are, and error messages are cryptic. I.e., compare:

```haskell
    <*> repeat (pure True)
    <*> repeat (pure True)
```

with

```haskell
    , txReadys = repeat (pure True) 
    , rxReadys = repeat (pure True) 
```

The latter is much more intelligible, I hope you agree. 

I intentionally implement this because I was getting tired of the bad error messages and counting arguments after introducing new ones. And yes, I wish we had something like anonymous records / row polymorphism to make this a whole bunch easier.

Reply:
I still don't get your point. If you don't like positional arguments then use `ApplicativeDo`, e.g.,
```haskell
transceivers =
  transceiverPrbs @GthTx @GthRx @Ext200 @Basic125 @GthTx @GthRx
    defConfig{debugIla=True, debugFpgaIndex=bitCoerce <$> fpgaIndex}
      <$> do transceiverIndex <- iterateI (+1) 0
             channelName <- channelNames
             clockPath <- clockPaths
             rxN <- unbundle (unpack <$> rxNs)
             rxP <- unbundle (unpack <$> rxPs)
             txData <- counters
             pure Input
               { clock = sysClk
               , reset = gthAllReset
               , refClock = refClk
               , txReady = pure True
               , rxReady = pure True
               , ..
               }
```
This way of writing offers the same readability as your approach (IMO), just without the necessity of all the overhead. And users that prefer using positional arguments are also not restricted in using them.

_The only flaw with this extension, I don't like, is that it does not support `in`-free `let`-bindings as the normal do notation does. You can use such let bindings, but GHC then always enforces the context to be monadic instead of applicative, which makes no sense somehow._

Reply:
I used `<$>`/`<*>` before but that triggers the ILA/VIO Clash bug mentioned elsewhere..

Other than the bug, I didn't think of using `ApplicativeDo` like this before, this is interesting! I'm still a bit on the fence, because you do push the boiler plate around at bit. But instead of doing it once, you have to do it at every call site:

```haskell
      <$> do transceiverIndex <- iterateI (+1) 0
             channelName <- channelNames
             clockPath <- clockPaths
             rxN <- unbundle (unpack <$> rxNs)
             rxP <- unbundle (unpack <$> rxPs)
             txData <- counters
```

I.e., for _N_ arguments you add _N_ lines of boilerplate on call sites. Not that that's necessarily bad, boilerplate is much less of a concern than hard-to-inspect code.

Initially I tried getting it to work using [barbies](https://hackage.haskell.org/package/barbies), preventing the overhead of having to define both `Input` and `Inputs`. Unfortunately, I never got it t work for reasons I've forgotten..

In any case, named arguments for functions this big are definitely the way to go, there is no doubt in my mind. (I even think named arguments is the biggest advantage Python has over Haskell in terms of readability!) Given that the `ApplicativeDo` approach is currently being blocked by Clash bugs, I think we should just leave it.